cli/azd/pkg/input/previewer.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package input
+
+import (
+	"strings"
+
+	tm "github.com/buger/goterm"
+)
+
+type previewer struct {
+	title        string
+	displayTitle string
+	footerLine   string
+	lines        int
+	prefix       string
+	output       []string
+}
+
+func NewPreviewer(lines int, prefix, title string) *previewer {
+	return &previewer{
+		lines:  lines,
+		prefix: prefix,
+		title:  title,
+	}
+}
+
+func (p *previewer) Start() {
+	p.output = make([]string, p.lines)
+	// title is created on Start() because it depends on terminal width
+	// if terminal is resized between stop and start, the previewer will
+	// react to it and update the size.
+	p.buildTitleMargins()
+	p.printOutput()
+}
+
+func clearLine(text string) {
+	sizeLog := len(text)
+	if sizeLog == 0 {
+		return
+	}
+	maxWidth := tm.Width()
+	if sizeLog > maxWidth {
+		sizeLog = maxWidth
+	}
+	eraseWith := strings.Repeat(" ", sizeLog)
+	tm.Printf(eraseWith)
+	tm.MoveCursorBackward(sizeLog)
+}
+
+func (p *previewer) Stop() {
+	p.clear()
+	p.output = nil
+}
+
+func (p *previewer) clear() {
+	if p.output == nil {
+		return
+	}
+
+	tm.MoveCursorBackward(len(p.footerLine))
+	clearLine(p.footerLine)
+	// skip empty line between footer and log
+	tm.MoveCursorUp(1)
+	size := len(p.output) - 1
+	for index := range p.output {
+		tm.MoveCursorUp(1)
+		clearLine(p.output[size-index])
+	}
+	// moving up two times, one is the empty line between title and logs
+	tm.MoveCursorUp(1)
+	tm.MoveCursorUp(1)
+	clearLine(p.displayTitle)
+
+	tm.Flush()
+}
+
+func (p *previewer) Write(logBytes []byte) (int, error) {
+	fullText := strings.Split(string(logBytes), "\n")
+	maxWidth := tm.Width()
+	if maxWidth <= 0 {
+		// tm.Width <= 0 means a CI terminal, where logs can be just written
+		return tm.Println(fullText)
+	}
+
+	for _, log := range fullText {
+		if log == "" {
+			continue
+		}
+		fullLog := p.prefix + log
+		if len(fullLog) > maxWidth {
+			fullLog = fullLog[:maxWidth-4] + cPostfix
+		}
+
+		p.clear()
+		p.output = pushRemove(p.output, fullLog)
+		p.printOutput()
+	}
+	return len(logBytes), nil
+}
+
+func pushRemove(original []string, value string) []string {
```

Comment:
It feels like inlining this, perhaps with a one or two sentence comment about what we are doing may make things overall more readable.  We wouldn't have to come up with a short, evocative name for this operation.  It's only called from one location.